### PR TITLE
ai_user cookie not present after reenabling the cookie #1585

### PR DESCRIPTION
**Pull request by MSNev**
- Add optimization to avoid always updating the cookie value

### CODE REVIEW

File: extensions/applicationinsights-properties-js/src/Context/User.ts [hunk 1]
```diff
@@ -75,6 +83,8 @@ export class User implements IUserContext {
                 const params = cookie.split(User.cookieSeparator);
                 if (params.length > 0) {
                     _self.id = params[0];
+                    // we already have a cookie
+                    _self.isUserCookieSet = !!_self.id;
```
Comment (MSNev) [inline extensions/applicationinsights-properties-js/src/Context/User.ts:87]: If we successfully read a value from the cookie then we assume that the cookie was set.

File: extensions/applicationinsights-properties-js/src/Context/User.ts [hunk 2]
```diff
@@ -98,7 +108,7 @@ export class User implements IUserContext {
                 // set it to 365 days from now
                 // 365 * 24 * 60 * 60 = 31536000 
                 const oneYear = 31536000;
-                _cookieManager.set(_storageNamePrefix(), cookie, oneYear);
+                _self.isUserCookieSet = _cookieManager.set(_storageNamePrefix(), cookie, oneYear);
```
Comment (MSNev) [inline extensions/applicationinsights-properties-js/src/Context/User.ts:111]: Use new response that tells us if the cookie was really set.

File: extensions/applicationinsights-properties-js/src/Context/User.ts [hunk 3]
```diff
@@ -172,9 +182,12 @@ export class User implements IUserContext {
             };
 
             _self.update = (userId?: string) => {
-                let user_id = userId ? userId : _generateNewId();
-                let user_cookie = _generateNewCookie(user_id);
-                _setUserCookie(user_cookie.join(User.cookieSeparator));
+                // Optimizations to avoid setting and processing the cookie when not needed
+                if (_self.id !== userId || !_self.isUserCookieSet) {
```
Comment (MSNev) [inline extensions/applicationinsights-properties-js/src/Context/User.ts:186]: Add an extra performance based check so we don't try and set the cookie if
- The value was already set
- The new userId is the same as the previous one.

File: extensions/applicationinsights-properties-js/src/PropertiesPlugin.ts [hunk 4]
```diff
@@ -92,14 +92,15 @@ export default class PropertiesPlugin extends BaseTelemetryPlugin implements IPr
...
+                    let userCtx = theContext.user;
+                    if (userCtx && !userCtx.isUserCookieSet) {
+                        userCtx.update(theContext.user.id);
                     }
     
                     _processTelemetryInternal(event, itemCtx);
     
-                    if (theContext.user && theContext.user.isNewUser) {
-                        theContext.user.isNewUser = false;
+                    if (userCtx && userCtx.isNewUser) {
```
Comment (MSNev) [inline extensions/applicationinsights-properties-js/src/PropertiesPlugin.ts:102]: Simple minification improvement which also has a miner perf improvement.

File: shared/AppInsightsCore/src/JavaScriptSDK.Interfaces/ICookieMgr.ts [hunk 5]
```diff
@@ -20,8 +20,9 @@ export interface ICookieMgr {
      * @param maxAgeSec - [optional] The maximum number of SECONDS that this cookie should survive
      * @param domain - [optional] The domain to set for the cookie
      * @param path - [optional] Path to set for the cookie, if not supplied will default to "/"
+     * @returns - True if the cookie was set otherwise false (Because cookie usage is not enabled or available)
      */
-    set(name: string, value: string, maxAgeSec?: number, domain?: string, path?: string): void;
+    set(name: string, value: string, maxAgeSec?: number, domain?: string, path?: string): boolean;
```
Comment (MSNev) [inline shared/AppInsightsCore/src/JavaScriptSDK.Interfaces/ICookieMgr.ts:25]: While this is a change to the signature of the function, it should be backward compatible as previously it was void.

Comment (MSNev) [reply shared/AppInsightsCore/src/JavaScriptSDK.Interfaces/ICookieMgr.ts:25]: This is also used by internally as well, so this should be compatible with that as well.

File: shared/AppInsightsCore/src/JavaScriptSDK/CookieMgr.ts [hunk 6]
```diff
@@ -190,6 +191,8 @@ export function createCookieMgr(rootConfig?: IConfiguration, logger?: IDiagnosti
                 let setCookieFn = cookieMgrConfig.setCookie || _setCookieValue;
                 setCookieFn(name, _formatCookieValue(theValue, values));
             }
+
+            return result;
```
Comment (xiao-lix) [inline shared/AppInsightsCore/src/JavaScriptSDK/CookieMgr.ts:195]: Do you forget to set result = true if cookie manager is enabled? Looks like the result will always be false.

Comment (MSNev) [reply shared/AppInsightsCore/src/JavaScriptSDK/CookieMgr.ts:195]: nice catch!

Comment (MSNev) [reply shared/AppInsightsCore/src/JavaScriptSDK/CookieMgr.ts:196]: Added

**Review by xiao-lix — APPROVED** [review]

**Pull request merged**